include/pinocchio/multibody/joint/joint-ellipsoid.hpp

```diff
@@ -0,0 +1,601 @@
+  struct JointDataEllipsoidTpl : public JointDataBase<JointDataEllipsoidTpl<_Scalar, _Options>>
+  {
+    EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+    typedef JointEllipsoidTpl<_Scalar, _Options> JointDerived;
+    PINOCCHIO_JOINT_DATA_TYPEDEF_TEMPLATE(JointDerived);
+    PINOCCHIO_JOINT_DATA_BASE_DEFAULT_ACCESSOR
+
+    ConfigVector_t joint_q;
+    TangentVector_t joint_v;
+
+    Constraint_t S;
+    Constraint_t Sdot;
+    Transformation_t M;
+    Motion_t v;
+    Bias_t c;
+
+    // [ABA] specific data
+    U_t U;
+    D_t Dinv;
+    UD_t UDinv;
+    D_t StU;
+
+    JointDataEllipsoidTpl()
+    : joint_q(ConfigVector_t::Zero())
+    , joint_v(TangentVector_t::Zero())
+    , S()
+    , Sdot()
+    , M(Transformation_t::Identity())
+    , v(Motion_t::Zero())
+    , c(Bias_t::Zero())
+    , U(U_t::Zero())
+    , Dinv(D_t::Zero())
+    , UDinv(UD_t::Zero())
+    , StU(D_t::Zero())
+    {
+    }
+
+    static std::string classname()
+    {
+      return std::string("JointDataEllipsoid");
+    }
+    std::string shortname() const
+    {
+      return classname();
+    }
+
+  }; // struct JointDataEllipsoidTpl
+
+  PINOCCHIO_JOINT_CAST_TYPE_SPECIALIZATION(JointModelEllipsoidTpl);
+  template<typename _Scalar, int _Options>
+  struct JointModelEllipsoidTpl : public JointModelBase<JointModelEllipsoidTpl<_Scalar, _Options>>
+  {
+    EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+    typedef JointEllipsoidTpl<_Scalar, _Options> JointDerived;
+    PINOCCHIO_JOINT_TYPEDEF_TEMPLATE(JointDerived);
+
+    typedef JointModelBase<JointModelEllipsoidTpl> Base;
+    using Base::id;
+    using Base::idx_q;
+    using Base::idx_v;
+    using Base::idx_vExtended;
+    using Base::setIndexes;
+
+    typedef typename Transformation_t::Vector3 Vector3;
+
+    Scalar radius_a;
+    Scalar radius_b;
+    Scalar radius_c;
+
+    JointDataDerived createData() const
+    {
+      return JointDataDerived();
+    }
+
+    JointModelEllipsoidTpl()
+    : radius_a(Scalar(0.01))
+    , radius_b(Scalar(0.01))
+    , radius_c(Scalar(0.01))
+    {
+    }
+
+    explicit JointModelEllipsoidTpl(const Scalar & a, const Scalar & b, const Scalar & c)
+    : radius_a(a)
+    , radius_b(b)
+    , radius_c(c)
+    {
+    }
+
+    const std::vector<bool> hasConfigurationLimit() const
+    {
+      return {true, true, true};
+    }
+
+    const std::vector<bool> hasConfigurationLimitInTangent() const
+    {
+      return {true, true, true};
+    }
+
+    template<typename ConfigVector>
+    void calc(JointDataDerived & data, const typename Eigen::MatrixBase<ConfigVector> & qs) const
+    {
+      data.joint_q = qs.template segment<NQ>(idx_q());
+
+      Scalar c0, s0;
+      SINCOS(data.joint_q(0), &s0, &c0);
+      Scalar c1, s1;
+      SINCOS(data.joint_q(1), &s1, &c1);
+      Scalar c2, s2;
+      SINCOS(data.joint_q(2), &s2, &c2);
+
+      data.M.rotation() << c1 * c2, -c1 * s2, s1, c0 * s2 + c2 * s0 * s1, c0 * c2 - s0 * s1 * s2,
+        -c1 * s0, -c0 * c2 * s1 + s0 * s2, c0 * s1 * s2 + c2 * s0, c0 * c1;
```

Comment:
You can deactivate clang-format to improve readability: see include/pinocchio/spatial/inertia.hpp:1008

```
      // clang-format off
      I_bar << dynamic_params[4], dynamic_params[5], dynamic_params[7],
               dynamic_params[5], dynamic_params[6], dynamic_params[8],
               dynamic_params[7], dynamic_params[8], dynamic_params[9];
      // clang-format on
```